### PR TITLE
[Messenger] Allow to typehint multiple buses

### PR DESCRIPTION
**Pull request by sroze**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ø
| License       | MIT
| Doc PR        | ø

When defining multiple buses, it becomes impossible to type-hint the `MessageBusInterface`. This PR adds the ability to wrap the bus within a decorator that you will be able to typehint.

As a developer, you will create your classes:
```php
namespace App\Bus;

use Symfony\Component\Messenger\DecoratedMessageBus;

final class CommandBus extends DecoratedMessageBus
{
}
```

Configure your bus to use it:
```yaml
framework:
    messenger:
        buses:
            commands:
                decorator_class: 'App\Bus\CommandBus'
```

You can now typehint your `CommandBus` 🥁 💃 ⭐️

### CODE REVIEW

File: src/Symfony/Component/Messenger/DecoratedMessageBus.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger;
+
+/**
+ * A decorated message bus.
+ *
+ * Use this abstract class to created your message bus decorator to specialise your
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/DecoratedMessageBus.php]: typo `to create`

File: src/Symfony/Component/Messenger/DecoratedMessageBus.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * bus instances and type-hint them.
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/DecoratedMessageBus.php]: I suggest `type-hint against them`, dependents type-hint one of their dependencies against the class

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 3]
```diff
@@ -1481,8 +1481,13 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
                 throw new LogicException('The Validation middleware is only available when the Validator component is installed and enabled. Try running "composer require symfony/validator".');
             }
 
+            $tagAttributes = array('name' => $name);
+            if (isset($bus['decorator_class'])) {
+                $tagAttributes['decorator_class'] = $bus['decorator_class'];
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: should it fail if the class does not implement `MessageBusInterface` at least? Perhaps give a complete hint about what is needed for using this option

File: src/Symfony/Component/Messenger/Tests/DependencyInjection/MessengerPassTest.php [hunk 4]
```diff
@@ -413,3 +425,7 @@ public function handle($message, callable $next)
         return $next($message);
     }
 }
+
+final class MyTypeHintedBus extends DecoratedMessageBus
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/Tests/DependencyInjection/MessengerPassTest.php]: TypeHinted is misleading here :) a class is a type, method arguments are type-hinted

**Review by chalasr — COMMENTED** [review]
mainly wording suggestions

**Comment by ro0NL** [comment]
>  it becomes impossible to type-hint the MessageBusInterface

you'll get the default bus AFAIK...

> You can now typehint your CommandBus

What if the same decorated class is used multiple times across buses?

isn't this just like saying `My\CommandBus: '@messenger.bus.command`, using regular DI wiring. Put different; do we really want/need this feature?

**Comment by sroze** [comment]
> What if the same decorated class is used multiple times across buses?

Well, you know what is going to happen: the auto-wiring won't be happy :)

> isn't this just like saying My\CommandBus: '@messenger.bus.command', using regular DI wiring. Put different; do we really want/need this feature?

Ish. It still means you'd have to create the `My\CommandBus` class, but it's a good point. Should we go with just the class and no FWB configuration?

**Comment by ro0NL** [comment]
Or just a trait maybe =/ i would indeed avoid the FWB config.

**Comment by sroze** [comment]
It works well with the following configuration 👍 
```yaml
    'App\Bus\CommandBus':
        decorates: "messenger.bus.commands"
```

**Comment by sroze** [comment]
To be completed by a lovely documentation :)

**Comment by kbond** [comment]
In development, with the following configuration:

```yaml
services:
    'App\Bus\CommandBus':
        decorates: "messenger.bus.commands"
```

The decorated bus is an instance of `Symfony\Component\Messenger\MessageBus` not `Symfony\Component\Messenger\TraceableMessageBus`

**Comment by sroze** [comment]
@kbond you need to typehint ither `MessageBusInterface` or your own `App\Bus\CommandBus` not `Symfony\Component\Messenger\MessageBus` 😉

**Comment by kbond** [comment]
@sroze I type-hinted my own bus which extends from `AbstractMessageBusDecorator`. The `$decoratedBus` property is not and instance of `Symfony\Component\Messenger\TraceableMessageBus`

**Comment by sroze** [comment]
I'm sorry but I don't understand the problem... "it works well on my machine" 😅 Can you give me the exact error you face and the classes & configuration you have? :)

**Comment by kbond** [comment]
There is no error.

messenger config:

```yaml
framework:
    messenger:
        default_bus: event
        buses:
            command:
                default_middlewares: false
                middlewares:
                    - logging
                    - Symfony\Bridge\Doctrine\Messenger\DoctrineTransactionMiddleware
                    - call_message_handler
            query:
                default_middlewares: false
                middlewares:
                    - logging
                    - call_message_handler
            event:
                middleware:
                    - tolerate_no_handler
```

service config:

```yaml
services:
    App\Messenger\CommandBus:
        decorates: messenger.bus.command
```

my command bus class:

```php
namespace App\Messenger;

use Symfony\Component\Messenger\AbstractMessageBusDecorator;

final class CommandBus extends AbstractMessageBusDecorator
{
}
```

controller injecting command bus:

```php
public function myAction(CommandBus $commandBus)
{
    dump($commandBus);

    // ...
}
```

Screenshot of dumped `$commandBus`:

symfony profiler

Notice the decorated bus is just `MessageBus` not `TraceableMessageBus`.

**Comment by kbond** [comment]
Note, if I change my service config to:

```yaml
services:
    App\Messenger\CommandBus: ['@messenger.bus.command']
```

symfony profiler 1

The `TraceableMessageBus` is the decoratedBus.

**Comment by sroze** [comment]
Status: Ready

**Review by ro0NL — APPROVED** [review]
not sure core needs it.. (i'd write this out anyway :)) but :+1: for the intend allthough i'd use a trait to avoid inheritance and own the private decorated bus in my own class.

**Review by weaverryan — APPROVED** [review]
Hmm, it seems like a pretty small feature in its final form. So, I think it's ok. @sroze can you open a Docs PR before this gets merged with a quick dump of the final "docs" needed to make this work? I'm not 100% clear which service configuration was finally the correct one from the comments.

**Comment by sroze** [comment]
@weaverryan https://github.com/symfony/symfony-docs/pull/9727

**Comment by nicolas-grekas** [comment]
I'm very skeptical about this PR and the related § in symfony/symfony-docs#9727.
This encourages type-hinting against a concrete implementation and totally breaks SOLID, isn't it?
I acknowledge there is a DX issue and that this proposal tries to solve it. Coincidentally, I was talking about the exact same issue with @weaverryan today, tell him about a solution we might want to push forward in 4.2.

**Comment by sroze** [comment]
As you know, I agree with you... What if we change the documentation to
show an example with creating the interface? For 4.1, this is very likely
to be the only option we can have in IMHO :)

**Comment by nicolas-grekas** [comment]
No need for any dedicated interfaces: using bindings solves the issues:
`$commandBus: @mesdenger.bus.commands`

**Comment by sroze** [comment]
Outch. Not sure how I feel about that one.

**Comment by nicolas-grekas** [comment]
As explained, I'm:-1: as this would put users in a dead end. We should tell about using bindings instead, a seek for a better experience in 4.2 that doesn't break SOLID.

**Comment by sroze** [comment]
Closing, as I'll update the documentation to recommend the binding, that makes sense actually.

**Pull request closed without merging**